File: src/main/python/leetcode/linkedlist/convert_sorted_list_to_binary_search_tree.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class RecursiveSolution:

    def findMiddleElement(self, head: ListNode) -> TreeNode:

        prev = None
        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            prev = slow
            slow = slow.next
            fast = fast.next.next

        if prev is not None:
            prev.next = None

        return slow

    def sortedListToBST(self, head: ListNode) -> TreeNode:
        if head is None:
            return None

        middle = self.findMiddleElement(head)

        node = TreeNode(middle.val)

        if head == middle:
            return node

        node.left = self.sortedListToBST(head)
        node.right = self.sortedListToBST(middle.next)

        return node

# class IterativeSolution:
# 	def sortedListToBST(self, head: ListNode) -> TreeNode:

File: src/main/go/linkedlist/convert_sorted_list_to_binary_search_tree.go
package linkedlist

func findMiddleElement(head *ListNode) *ListNode {
	var prev *ListNode
	fast := head
	slow := head

	for fast != nil && fast.Next != nil {
		prev = slow
		slow = slow.Next
		fast = fast.Next.Next
	}

	if prev != nil {
		prev.Next = nil
	}
	return slow
}

func sortedListToBST(head *ListNode) *TreeNode {
	if head == nil {
		return nil
	}

	middle := findMiddleElement(head)

	node := TreeNode{
		Val: middle.Val,
	}

	node.Left = sortedListToBST(middle)
	node.Right = sortedListToBST(middle.Next)

	return &node
}

func sortedListToBSTIteratively(head *ListNode) *TreeNode {
	array := getArray(head)
	l := len(array)

	return sortedListToBSTIterativelyHelper(0, l-1, array)
}

func getArray(head *ListNode) []*ListNode {
	var array []*ListNode

	for head != nil {
		array = append(array, head)
		head = head.Next
	}

	return array
}

func sortedListToBSTIterativelyHelper(left int, right int, nodes []*ListNode) *TreeNode {

	if left < right {
		return nil
	}

	middle := left + right/2

	node := TreeNode{
		Val: nodes[middle].Val,
	}

	node.Left = sortedListToBSTIterativelyHelper(left, middle-1, nodes)
	node.Right = sortedListToBSTIterativelyHelper(middle+1, right, nodes)

	return &node
}
